File: example/language/hello.ex
# Elixir 示例文件
# 展示 Elixir 语言的各种语法特性

defmodule HelloElixir do
  # 模块属性
  @app_name "Elixir Example"
  @version "1.0.0"
  @pi 3.14159

  # 模块常量
  defmacro app_name, do: @app_name
  defmacro version, do: @version
  defmacro pi, do: @pi

  # 全局状态（使用Agent）
  def start_link do
    Agent.start_link(fn -> %{counter: 0} end, name: __MODULE__)
  end

  def get_counter do
    Agent.get(__MODULE__, & &1.counter)
  end

  def increment_counter do
    Agent.update(__MODULE__, & %{&1 | counter: &1.counter + 1})
  end
end

# 结构体定义
defmodule Person do
  defstruct [:name, :age, :email, :id]

  # 构造函数
  def new(name, age, email \\ nil) do
    HelloElixir.increment_counter()
    %Person{
      name: name,
      age: age,
      email: email,
      id: HelloElixir.get_counter()
    }
  end

  # 方法
  def greet(%Person{name: name, age: age}) do
    "Hello, my name is #{name} and I'm #{age} years old!"
  end

  def adult?(%Person{age: age}) do
    age >= 18
  end

  def celebrate_birthday(%Person{} = person) do
    IO.puts("Happy birthday! You're now #{person.age + 1} years old.")
    %{person | age: person.age + 1}
  end
end

# 协议（Protocol）定义
defprotocol Shape do
  def area(shape)
  def perimeter(shape)
  def description(shape)
end

# 具体实现
defmodule Circle do
  defstruct [:radius]

  defimpl Shape do
    def area(%Circle{radius: r}), do: @pi * r * r
    def perimeter(%Circle{radius: r}), do: 2 * @pi * r
    def description(%Circle{radius: r}), do: "Circle with radius #{Float.round(r, 2)}"
  end
end

defmodule Rectangle do
  defstruct [:width, :height]

  defimpl Shape do
    def area(%Rectangle{width: w, height: h}), do: w * h
    def perimeter(%Rectangle{width: w, height: h}), do: 2 * (w + h)
    def description(%Rectangle{width: w, height: h}), do: "Rectangle #{Float.round(w, 2)} x #{Float.round(h, 2)}"
  end
end

defmodule Triangle do
  defstruct [:a, :b, :c]

  defimpl Shape do
    def area(%Triangle{a: a, b: b, c: c}) do
      s = (a + b + c) / 2
      :math.sqrt(s * (s - a) * (s - b) * (s - c))
    end

    def perimeter(%Triangle{a: a, b: b, c: c}), do: a + b + c

    def description(%Triangle{a: a, b: b, c: c}) do
      "Triangle with sides #{Float.round(a, 2)}, #{Float.round(b, 2)}, #{Float.round(c, 2)}"
    end
  end
end

# 泛型容器模块
defmodule Container do
  defstruct [:items]

  def new, do: %Container{items: []}

  def add(%Container{items: items} = container, item) do
    %{container | items: [item | items]}
  end

  def remove(%Container{items: items} = container, index) when index >= 0 and index < length(items) do
    {removed, rest} = List.pop_at(items, index)
    {%{container | items: rest}, removed}
  end

  def remove(container, _index), do: {container, nil}

  def get(%Container{items: items}, index) do
    Enum.at(items, index)
  end

  def size(%Container{items: items}), do: length(items)

  def map(%Container{items: items}, func), do: Enum.map(items, func)

  def filter(%Container{items: items}, predicate), do: Enum.filter(items, predicate)

  def reduce(%Container{items: items}, acc, func), do: Enum.reduce(items, acc, func)
end

# 递归函数
defmodule MathUtils do
  def factorial(0), do: 1
  def factorial(1), do: 1
  def factorial(n) when n > 1, do: n * factorial(n - 1)

  def fibonacci(0), do: []
  def fibonacci(1), do: [0]
  def fibonacci(n) when n > 1 do
    fib_helper([0, 1], n - 2)
  end

  defp fib_helper(sequence, 0), do: Enum.reverse(sequence)
  defp fib_helper([a, b | _] = sequence, count) do
    fib_helper([b + a, a, b | tl(sequence)], count - 1)
  end
end

# 高阶函数和函数式编程
defmodule FunctionalUtils do
  def process_list(list, func), do: Enum.map(list, func)
  def filter_list(list, predicate), do: Enum.filter(list, predicate)
  def reduce_list(list, acc, func), do: Enum.reduce(list, acc, func)

  def compose(f, g) do
    fn x -> f.(g.(x)) end
  end
end

# 宏定义
defmodule Macros do
  defmacro unless(condition, clauses) do
    quote do
      if !unquote(condition) do
        unquote(clauses)
      end
    end
  end

  defmacro measure_time(block) do
    quote do
      start = System.monotonic_time(:microsecond)
      result = unquote(block)
      finish = System.monotonic_time(:microsecond)
      {result, finish - start}
    end
  end
end

# 错误处理
defmodule Validation do
  defmodule ValidationError do
    defexception [:message, :field]
  end

  def validate_email(email) do
    email_regex = ~r/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/

    if Regex.match?(email_regex, email) do
      :ok
    else
      {:error, %ValidationError{message: "Invalid email format: #{email}", field: :email}}
    end
  end

  def validate_age(age) when age < 0 do
    {:error, %ValidationError{message: "Age cannot be negative", field: :age}}
  end

  def validate_age(age) when age > 150 do
    {:error, %ValidationError{message: "Age cannot be greater than 150", field: :age}}
  end

  def validate_age(_age), do: :ok

  def safe_divide(_a, 0), do: {:error, "Division by zero"}
  def safe_divide(a, b), do: {:ok, a / b}
end

# 字符串处理
defmodule StringUtils do
  def extract_numbers(text) do
    Regex.scan(~r/\d+/, text)
    |> Enum.map(&String.to_integer/1)
  end

  def capitalize_words(text) do
    text
    |> String.split()
    |> Enum.map(&String.capitalize/1)
    |> Enum.join(" ")
  end
end

# 并发示例
defmodule Concurrent do
  def process_numbers(numbers) do
    numbers
    |> Enum.map(&Task.async(fn -> process_number(&1) end))
    |> Enum.map(&Task.await/1)
  end

  defp process_number(n) do
    Process.sleep(100) # 模拟耗时操作
    n * n
  end
end

# 文件操作
defmodule FileUtils do
  def read_lines(filename) do
    case File.read(filename) do
      {:ok, content} ->
        content
        |> String.split("\n")
        |> Enum.reject(&(&1 == ""))

      {:error, reason} ->
        IO.puts("Error reading file: #{reason}")
        []
    end
  end

  def write_lines(filename, lines) do
    content = Enum.join(lines, "\n")

    case File.write(filename, content) do
      :ok -> IO.puts("File written: #{filename}")
      {:error, reason} -> IO.puts("Error writing file: #{reason}")
    end
  end
end

# 管道操作符示例
defmodule PipelineExample do
  def process_data(data) do
    data
    |> Enum.filter(&is_number/1)
    |> Enum.map(&(&1 * 2))
    |> Enum.sum()
  end
end

# 模式匹配和卫兵子句
defmodule PatternMatching do
  def describe_number(n) when n < 0, do: "negative"
  def describe_number(0), do: "zero"
  def describe_number(n) when n < 10, do: "small positive"
  def describe_number(n) when rem(n, 2) == 0, do: "even"
  def describe_number(_n), do: "other"

  def process_tuple({:ok, value}), do: "Success: #{value}"
  def process_tuple({:error, reason}), do: "Error: #{reason}"
  def process_tuple(_), do: "Unknown result"
end

# 主模块
defmodule Main do
  import Macros

  def run do
    IO.puts("=== Elixir Example ===\n")

    # 启动全局状态
    HelloElixir.start_link()

    # 基本数据类型
    IO.puts("--- Basic Data Types ---")
    str = "Hello, Elixir World!"
    num = 42
    float_num = 3.14159
    bool = true
    atom = :example
    nil_val = nil

    IO.puts("String: #{str}")
    IO.puts("Number: #{num}")
    IO.puts("Float: #{float_num}")
    IO.puts("Boolean: #{bool}")
    IO.puts("Atom: #{atom}")
    IO.puts("Nil: #{nil_val}")
    IO.puts("")

    # 集合操作
    IO.puts("--- Collections ---")
    numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    even_numbers = Enum.filter(numbers, &Integer.is_even/1)
    squared_numbers = Enum.map(numbers, &(&1 * &1))

    IO.puts("Original: #{inspect(numbers)}")
    IO.puts("Even numbers: #{inspect(even_numbers)}")
    IO.puts("Squared: #{inspect(squared_numbers)}")
    IO.printf("Sum: %d, Average: %.2f%n", Enum.sum(numbers), Enum.sum(numbers) / length(numbers))
    IO.puts("")

    # 结构体使用
    IO.puts("--- Structs ---")
    person1 = Person.new("Alice", 30, "alice@example.com")
    person2 = Person.new("Bob", 25)

    IO.puts("Person 1: #{inspect(person1)}")
    IO.puts("Person 2: #{inspect(person2)}")
    IO.puts("Person 1 greeting: #{Person.greet(person1)}")
    IO.puts("Is person 1 adult? #{Person.adult?(person1)}")

    older_person1 = Person.celebrate_birthday(person1)
    IO.puts("After birthday: #{inspect(older_person1)}")
    IO.puts("")

    # 形状示例
    IO.puts("--- Shapes ---")
    shapes = [
      %Circle{radius: 5.0},
      %Rectangle{width: 4.0, height: 6.0},
      %Triangle{a: 3.0, b: 4.0, c: 5.0}
    ]

    Enum.each(shapes, fn shape ->
      IO.printf("%s -> Area: %.2f, Perimeter: %.2f%n",
                Shape.description(shape), Shape.area(shape), Shape.perimeter(shape))
    end)
    IO.puts("")

    # 泛型容器
    IO.puts("--- Generic Container ---")
    container = Container.new()
    container = Container.add(container, "Hello")
    container = Container.add(container, "World")
    container = Container.add(container, "Elixir")

    IO.puts("Container size: #{Container.size(container)}")
    IO.puts("Container contents:")
    Enum.with_index(Container.map(container, & &1), fn item, index ->
      IO.puts("  #{index}: #{item}")
    end)

    long_words = Container.filter(container, &(String.length(&1) > 4))
    IO.puts("Long words: #{inspect(long_words)}")

    lengths = Container.map(container, &String.length/1)
    IO.puts("Word lengths: #{inspect(lengths)}")
    IO.puts("")

    # 函数式编程
    IO.puts("--- Functional Programming ---")
    doubled = FunctionalUtils.process_list(numbers, &(&1 * 2))
    greater_than_5 = FunctionalUtils.filter_list(numbers, &(&1 > 5))

    IO.puts("Original: #{inspect(numbers)}")
    IO.puts("Doubled: #{inspect(doubled)}")
    IO.puts("Greater than 5: #{inspect(greater_than_5)}")

    # 函数组合
    add_5 = &(&1 + 5)
    multiply_by_2 = &(&1 * 2)
    composed = FunctionalUtils.compose(multiply_by_2, add_5)
    IO.puts("Composed function result (multiply_by_2(add_5(10))): #{composed.(10)}")
    IO.puts("")

    # 递归和数学
    IO.puts("--- Recursion and Math ---")
    IO.puts("Factorial of 5: #{MathUtils.factorial(5)}")
    IO.puts("Fibonacci sequence: #{inspect(MathUtils.fibonacci(10))}")
    IO.puts("")

    # 宏使用
    IO.puts("--- Macros ---")
    unless false do
      IO.puts("This should print (unless with false)")
    end

    unless true do
      IO.puts("This should not print (unless with true)")
    end

    {result, time} = measure_time do
      Process.sleep(100)
      MathUtils.fibonacci(15)
    end

    IO.puts("Fibonacci calculation took #{time} microseconds")
    IO.puts("")

    # 字符串处理
    IO.puts("--- String Processing ---")
    text = "The numbers are 123, 456 and 789 in this text."
    extracted_numbers = StringUtils.extract_numbers(text)

    IO.puts("Original text: #{text}")
    IO.puts("Extracted numbers: #{inspect(extracted_numbers)}")
    IO.puts("Contains 'numbers': #{String.contains?(text, "numbers")}")
    IO.puts("Uppercase: #{String.upcase(text)}")
    IO.puts("Capitalized words: #{StringUtils.capitalize_words(text)}")
    IO.puts("")

    # 错误处理
    IO.puts("--- Error Handling ---")
    emails = ["valid@example.com", "invalid-email", "another@valid.com"]

    Enum.each(emails, fn email ->
      case Validation.validate_email(email) do
        :ok -> IO.puts("#{email}: Valid")
        {:error, error} -> IO.puts("#{email}: Error - #{error.message}")
      end
    end)

    case Validation.validate_age(-5) do
      :ok -> IO.puts("Age validation passed")
      {:error, error} -> IO.puts("Age validation error: #{error.message}")
    end

    case Validation.safe_divide(10, 2) do
      {:ok, result} -> IO.puts("10 / 2 = #{result}")
      {:error, reason} -> IO.puts("Division error: #{reason}")
    end

    case Validation.safe_divide(10, 0) do
      {:ok, result} -> IO.puts("10 / 0 = #{result}")
      {:error, reason} -> IO.puts("Division error: #{reason}")
    end
    IO.puts("")

    # Map 和 Keyword
    IO.puts("--- Maps and Keywords ---")
    grades = %{"Alice" => 95, "Bob" => 87, "Charlie" => 92, "Diana" => 88}

    IO.puts("Student grades:")
    Enum.each(grades, fn {name, grade} ->
      IO.puts("  #{name}: #{grade}")
    end)

    # 按成绩排序
    sorted_grades = Enum.sort_by(grades, fn {_name, grade} -> grade end, &>=/2)
    IO.puts("Sorted by grade (descending):")
    Enum.each(sorted_grades, fn {name, grade} ->
      IO.puts("  #{name}: #{grade}")
    end)
    IO.puts("")

    # 并发
    IO.puts("--- Concurrency ---")
    concurrent_results = Concurrent.process_numbers([1, 2, 3, 4, 5])
    IO.puts("Concurrent squares: #{inspect(Enum.sort(concurrent_results))}")
    IO.puts("")

    # 模式匹配
    IO.puts("--- Pattern Matching ---")
    test_numbers = [-5, 0, 1, 2, 4, 15, 100]

    Enum.each(test_numbers, fn n ->
      IO.puts("#{n} is #{PatternMatching.describe_number(n)}")
    end)

    # 元组模式匹配
    results = [{:ok, "success"}, {:error, "failed"}, {:unknown, "maybe"}]
    Enum.each(results, fn result ->
      IO.puts(PatternMatching.process_tuple(result))
    end)
    IO.puts("")

    # 管道操作符
    IO.puts("--- Pipeline Operator ---")
    mixed_data = [1, "hello", 3.14, 2, "world", 5]
    pipeline_result = PipelineExample.process_data(mixed_data)
    IO.puts("Pipeline result (sum of doubled numbers): #{pipeline_result}")
    IO.puts("")

    # 文件操作演示
    IO.puts("--- File Operations (Demo) ---")
    temp_content = [
      "This is a test file created by Elixir example.",
      "Line 2",
      "Line 3"
    ]

    FileUtils.write_lines("temp_elixir_example.txt", temp_content)

    lines = FileUtils.read_lines("temp_elixir_example.txt")
    IO.puts("File content:")
    Enum.with_index(lines, 1, fn line, index ->
      IO.puts("  #{index}: #{line}")
    end)

    # 清理
    File.rm("temp_elixir_example.txt")
    IO.puts("Cleaned up temporary file")
    IO.puts("")

    # 列表推导式
    IO.puts("--- List Comprehensions ---")
    squares = for n <- 1..5, do: n * n
    IO.puts("Squares of 1..5: #{inspect(squares)}")

    even_squares = for n <- 1..10, Integer.is_even(n), do: n * n
    IO.puts("Squares of even numbers 1..10: #{inspect(even_squares)}")

    coordinates = for x <- 1..2, y <- 1..2, do: {x, y}
    IO.puts("Coordinates: #{inspect(coordinates)}")
    IO.puts("")

    # 日期和时间
    IO.puts("--- Date and Time ---")
    now = DateTime.utc_now()
    IO.puts("Current time: #{now}")
    IO.puts("ISO format: #{DateTime.to_iso8601(now)}")
    IO.puts("Unix timestamp: #{DateTime.to_unix(now)}")
    IO.puts("")

    IO.puts("=== Elixir Example Completed ===")
  end
end

# 脚本入口
Main.run()

File: example/language/hello.clj
;; Clojure 示例文件
;; 展示 Clojure 语言的各种语法特性

(ns hello-clojure
  (:require [clojure.string :as str]
            [clojure.set :as set]))

;; 全局常量
(def APP-NAME "Clojure Example")
(def VERSION "1.0.0")
(def PI 3.14159)

;; 全局变量
(def global-counter (atom 0))

;; 记录（Record）定义
(defrecord Person [name age email id])

;; 构造函数
(defn make-person
  ([name age] (make-person name age nil))
  ([name age email]
   (swap! global-counter inc)
   (->Person name age email @global-counter)))

;; 协议（Protocol）定义
(defprotocol Shape
  (area [this])
  (perimeter [this])
  (description [this]))

;; 类型扩展
(defrecord Circle [radius]
  Shape
  (area [_] (* Math/PI radius radius))
  (perimeter [_] (* 2 Math/PI radius))
  (description [_] (format "Circle with radius %.2f" (double radius))))

(defrecord Rectangle [width height]
  Shape
  (area [_] (* width height))
  (perimeter [_] (* 2 (+ width height)))
  (description [_] (format "Rectangle %.2f x %.2f" (double width) (double height))))

(defrecord Triangle [a b c]
  Shape
  (area [_]
    (let [s (/ (+ a b c) 2)]
      (Math/sqrt (* s (- s a) (- s b) (- s c)))))
  (perimeter [_] (+ a b c))
  (description [_] (format "Triangle with sides %.2f, %.2f, %.2f"
                          (double a) (double b) (double c))))

;; 多重方法
(defmulti greet :type)

(defmethod greet :person [person]
  (format "Hello, my name is %s and I'm %d years old!"
          (:name person) (:age person)))

(defmethod greet :animal [animal]
  (format "Hello, I'm a %s!" (:species animal)))

;; Person 的方法
(defn person-greet [person]
  (format "Hello, my name is %s and I'm %d years old!"
          (:name person) (:age person)))

(defn person-adult? [person]
  (>= (:age person) 18))

(defn person-celebrate-birthday [person]
  (println (format "Happy birthday! You're now %d years old." (inc (:age person))))
  (update person :age inc))

;; 递归函数
(defn factorial [n]
  (if (<= n 1)
    1
    (* n (factorial (dec n)))))

(defn fibonacci [n]
  (cond
    (<= n 0) []
    (= n 1) [0]
    :else (loop [a 0 b 1 count 2 result [0 1]]
            (if (>= count n)
              result
              (recur b (+ a b) (inc count) (conj result (+ a b)))))))

;; 高阶函数
(defn process-collection [coll f]
  (map f coll))

(defn filter-collection [coll pred]
  (filter pred coll))

(defn reduce-collection [coll f initial]
  (reduce f initial coll))

;; 宏定义
(defmacro unless [test & body]
  `(when (not ~test)
     ~@body))

(defmacro with-timeout [timeout-ms & body]
  `(let [future# (future (do ~@body))
         timeout# (Thread/sleep ~timeout-ms)]
     (or (deref future# ~timeout-ms :timeout)
         (do (future-cancel future#)
             :timeout))))

;; 延迟求值
(def lazy-fibonacci
  (lazy-seq
   (cons 0
         (cons 1
               (map + lazy-fibonacci (rest lazy-fibonacci))))))

;; 工具函数
(defn safe-divide [a b]
  (if (zero? b)
    (throw (ArithmeticException. "Division by zero"))
    (/ a b)))

(defn validate-email [email]
  (let [email-regex #"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"]
    (if (re-matches email-regex email)
      true
      (throw (IllegalArgumentException. (str "Invalid email: " email))))))

(defn validate-age [age]
  (cond
    (neg? age) (throw (IllegalArgumentException. "Age cannot be negative"))
    (> age 150) (throw (IllegalArgumentException. "Age cannot be greater than 150"))
    :else true))

;; 集合操作
(defn analyze-numbers [numbers]
  {:original numbers
   :even (filter even? numbers)
   :odd (filter odd? numbers)
   :sum (reduce + numbers)
   :average (/ (reduce + numbers) (count numbers))
   :max (apply max numbers)
   :min (apply min numbers)})

;; 字符串处理
(defn extract-numbers [text]
  (let [number-regex #"\d+"]
    (->> (re-seq number-regex text)
         (map #(Integer/parseInt %)))))

;; 文件操作
(defn read-file-lines [filename]
  (try
    (with-open [reader (clojure.java.io/reader filename)]
      (doall (line-seq reader)))
    (catch Exception e
      (println (str "Error reading file: " (.getMessage e)))
      [])))

(defn write-to-file [filename lines]
  (try
    (with-open [writer (clojure.java.io/writer filename)]
      (doseq [line lines]
        (.write writer (str line "\n"))))
    (println (str "File written: " filename))
    (catch Exception e
      (println (str "Error writing file: " (.getMessage e))))))

;; 并发示例
(defn process-numbers-concurrent [numbers]
  (let [results (atom [])]
    (doseq [num numbers]
      (future
        (Thread/sleep 100) ; 模拟耗时操作
        (swap! results conj (* num num))))
    (Thread/sleep 200) ; 等待所有future完成
    @results))

;; 状态管理
(def person-store (atom {}))

(defn add-person [person]
  (swap! person-store assoc (:id person) person))

(defn get-person [id]
  (get @person-store id))

(defn list-people []
  (vals @person-store))

;; DSL 示例
(defmacro define-shape [name & {:keys [area perimeter description]}]
  `(defrecord ~name []
     Shape
     ~@(when area `((area [_] ~area)))
     ~@(when perimeter `((perimeter [_] ~perimeter)))
     ~@(when description `((description [_] ~description)))))

;; 主函数
(defn -main []
  (println "=== Clojure Example ===\n")

  ;; 基本数据类型
  (println "--- Basic Data Types ---")
  (def str-var "Hello, Clojure World!")
  (def num-var 42)
  (def float-var 3.14159)
  (def bool-var true)
  (def nil-var nil)

  (println (str "String: " str-var))
  (println (str "Number: " num-var))
  (println (str "Float: " float-var))
  (println (str "Boolean: " bool-var))
  (println (str "Nil: " nil-var))
  (println)

  ;; 集合操作
  (println "--- Collections ---")
  (def numbers [1 2 3 4 5 6 7 8 9 10])
  (def even-numbers (filter even? numbers))
  (def squared-numbers (map #(* % %) numbers))

  (println (str "Original: " numbers))
  (println (str "Even numbers: " (vec even-numbers)))
  (println (str "Squared: " (vec squared-numbers)))
  (printf "Sum: %d, Average: %.2f%n"
          (reduce + numbers)
          (/ (reduce + numbers) (count numbers)))
  (println)

  ;; Record 使用
  (println "--- Records ---")
  (def person1 (make-person "Alice" 30 "alice@example.com"))
  (def person2 (make-person "Bob" 25))

  (println (str "Person 1: " person1))
  (println (str "Person 2: " person2))
  (println (str "Person 1 greeting: " (person-greet person1)))
  (println (str "Is person 1 adult? " (person-adult? person1)))

  (def older-person1 (person-celebrate-birthday person1))
  (println (str "After birthday: " older-person1))
  (println)

  ;; 形状示例
  (println "--- Shapes ---")
  (def shapes [(->Circle 5.0) (->Rectangle 4.0 6.0) (->Triangle 3.0 4.0 5.0)])

  (doseq [shape shapes]
    (printf "%s -> Area: %.2f, Perimeter: %.2f%n"
            (description shape) (area shape) (perimeter shape)))
  (println)

  ;; 函数式编程
  (println "--- Functional Programming ---")
  (def doubled (process-collection numbers #(* 2 %)))
  (def greater-than-5 (filter-collection numbers #(> % 5)))

  (println (str "Original: " numbers))
  (println (str "Doubled: " (vec doubled)))
  (println (str "Greater than 5: " (vec greater-than-5)))
  (println)

  ;; 递归和数学
  (println "--- Recursion and Math ---")
  (println (str "Factorial of 5: " (factorial 5)))
  (println (str "Fibonacci sequence: " (fibonacci 10)))
  (println (str "Lazy Fibonacci (first 10): " (take 10 lazy-fibonacci)))
  (println)

  ;; 字符串处理
  (println "--- String Processing ---")
  (def text "The numbers are 123, 456 and 789 in this text.")
  (def extracted-numbers (extract-numbers text))

  (println (str "Original text: " text))
  (println (str "Extracted numbers: " extracted-numbers))
  (println (str "Contains 'numbers': " (str/includes? text "numbers")))
  (println (str "Uppercase: " (str/upper-case text)))
  (println)

  ;; 错误处理
  (println "--- Error Handling ---")
  (def emails ["valid@example.com" "invalid-email" "another@valid.com"])

  (doseq [email emails]
    (try
      (validate-email email)
      (println (str email ": Valid"))
      (catch IllegalArgumentException e
        (println (str email ": Error - " (.getMessage e))))))

  (try
    (validate-age -5)
    (catch IllegalArgumentException e
      (println (str "Age validation error: " (.getMessage e)))))
  (println)

  ;; Map 和 Set
  (println "--- Maps and Sets ---")
  (def grades {"Alice" 95, "Bob" 87, "Charlie" 92, "Diana" 88})

  (println "Student grades:")
  (doseq [[name grade] grades]
    (println (str "  " name ": " grade)))

  ;; 按成绩排序
  (def sorted-grades (sort-by val > grades))
  (println "Sorted by grade (descending):")
  (doseq [[name grade] sorted-grades]
    (println (str "  " name ": " grade)))
  (println)

  ;; 宏使用
  (println "--- Macros ---")
  (unless false
    (println "This should print (unless with false)"))

  (unless true
    (println "This should not print (unless with true)"))
  (println)

  ;; 并发
  (println "--- Concurrency ---")
  (def concurrent-results (process-numbers-concurrent [1 2 3 4 5]))
  (println (str "Concurrent squares: " (sort concurrent-results)))
  (println)

  ;; 集合分析
  (println "--- Collection Analysis ---")
  (def analysis (analyze-numbers numbers))
  (println "Number analysis:")
  (doseq [[key value] analysis]
    (println (str "  " (name key) ": " value)))
  (println)

  ;; 文件操作演示
  (println "--- File Operations (Demo) ---")
  (def temp-content ["This is a test file created by Clojure example."
                    "Line 2"
                    "Line 3"])

  (write-to-file "temp_clojure_example.txt" temp-content)

  (def lines (read-file-lines "temp_clojure_example.txt"))
  (println "File content:")
  (doseq [[i line] (map-indexed vector lines)]
    (println (str "  " (inc i) ": " line)))

  ;; 清理
  (clojure.java.io/delete-file "temp_clojure_example.txt")
  (println "Cleaned up temporary file")
  (println)

  ;; 状态管理
  (println "--- State Management ---")
  (add-person person1)
  (add-person person2)
  (println (str "People in store: " (count (list-people))))
  (println (str "Person with ID 1: " (get-person 1)))
  (println)

  ;; 序列操作
  (println "--- Sequence Operations ---")
  (def mixed-seq [1 "hello" 3.14 true nil])
  (println (str "Mixed sequence: " mixed-seq))
  (println (str "Types: " (map type mixed-seq)))
  (println (str "Non-nil values: " (filter some? mixed-seq)))
  (println)

  (println "=== Clojure Example Completed ==="))

;; 如果直接运行此文件
(when (= *file* (System/getProperty "babashka.file"))
  (-main))

;; 导出函数（如果作为库使用）
(def exports
  {:factorial factorial
   :fibonacci fibonacci
   :make-person make-person
   :person-greet person-greet})
